network/p2p/p2p.go

```diff
@@ -172,74 +176,96 @@ func New(ctx context.Context, logger *zap.Logger, cfg *Config) (network.Network,
+	}
+	n.cfg.Topics[validatorPk.SerializeToHexStr()] = topic
 
-			n.cfg.Sub.Cancel()
-		default:
-			msg, err := n.cfg.Sub.Next(n.ctx)
-			if err != nil {
-				n.logger.Error("failed to get message from subscription Topic", zap.Error(err))
-				return
-			}
+	sub, err := topic.Subscribe()
+	if err != nil {
+		return errors.Wrap(err, "failed to subscribe on Topic")
+	}
+	n.cfg.Subs = append(n.cfg.Subs, sub)
 
-			var cm network.Message
-			if err := json.Unmarshal(msg.Data, &cm); err != nil {
-				n.logger.Error("failed to unmarshal message", zap.Error(err))
-				continue
-			}
+	// listen to new topic
+	n.listen(sub)
+	return nil
+}
 
-			for _, ls := range n.listeners {
-				go func(ls listener) {
-					switch cm.Type {
-					case network.NetworkMsg_IBFTType:
-						ls.msgCh <- cm.SignedMessage
-					case network.NetworkMsg_SignatureType:
-						ls.sigCh <- cm.SignedMessage
-					case network.NetworkMsg_DecidedType:
-						ls.decidedCh <- cm.SignedMessage
-					default:
-						n.logger.Error("received unsupported message", zap.Int32("msg type", int32(cm.Type)))
-					}
-				}(ls)
+// ReceivedMsgChan return a channel with messages
+func (n *p2pNetwork) listen(sub *pubsub.Subscription) {
+	go func(sub *pubsub.Subscription) {
+		n.logger.Info("start listen to topic", zap.String("topic", sub.Topic()))
+
+		for {
+			select {
+			case <-n.ctx.Done():
+				if err := n.cfg.Topics[sub.Topic()].Close(); err != nil {
+					n.logger.Error("failed to close Topics", zap.Error(err))
+				}
+
+				sub.Cancel()
+			default:
+				msg, err := sub.Next(n.ctx)
+				if err != nil {
+					n.logger.Error("failed to get message from subscription Topics", zap.Error(err))
+					return
+				}
+
+				var cm network.Message
+				if err := json.Unmarshal(msg.Data, &cm); err != nil {
+					n.logger.Error("failed to unmarshal message", zap.Error(err))
+					continue
+				}
+
+				for _, ls := range n.listeners {
+					go func(ls listener) {
+						switch cm.Type {
+						case network.NetworkMsg_IBFTType:
+							ls.msgCh <- cm.SignedMessage
+						case network.NetworkMsg_SignatureType:
+							ls.sigCh <- cm.SignedMessage
+						case network.NetworkMsg_DecidedType:
+							ls.decidedCh <- cm.SignedMessage
+						default:
+							n.logger.Error("received unsupported message", zap.Int32("msg type", int32(cm.Type)))
+						}
+					}(ls)
+				}
 			}
 		}
+	}(sub)
+}
+
+func (n *p2pNetwork) GetTopic(msg *proto.SignedMessage) (*pubsub.Topic, error) {
+	if msg.Message == nil || msg.Message.ValidatorPk == nil{
+		return nil, errors.New("missing Message or ValidatorPk in signMsg")
+	}
+	pk := &bls.PublicKey{}
+	if err := pk.Deserialize(msg.Message.GetValidatorPk()); err != nil{
+		return nil, errors.Wrap(err, "failed to deserialize publicKey")
 	}
+	topic := pk.SerializeToHexStr()
+	if _, ok := n.cfg.Topics[topic]; !ok {
+		return nil, errors.New("topic is not exist or registered")
+	}
+	return n.cfg.Topics[topic], nil
 }
 
+
 func getTopic(topicName string) string {
```

Comment:
@nivBlox we need this function?

Reply:
Yes, each validator need to fetch the proper topic in order to broadcast msg